06-classification_continued.Rmd

```diff
@@ -559,16 +559,257 @@ The overall workflow for performing K-nearest neighbour classification using `ti
 
 **Strengths:**
 
-1. Simple and easy to understand
-2. No assumptions about what the data must look like 
-3. Works easily for binary (two-class) and multi-class (> 2 classes) classification problems
+1. Simple, intuitive algorithm
+2. Few assumptions about what the data must look like 
+3. Works for binary (two-class) and multi-class (> 2 classes) classification problems
 
 **Weaknesses:**
 
-1. As data gets bigger and bigger, K-nearest neighbour gets slower and slower, quite quickly
-2. Does not perform well with a large number of predictors
+1. As the training data gets larger, K-nearest neighbours becomes very slow
+2. May not perform well with a large number of predictors
 3. Does not perform well when classes are imbalanced (when many more observations are in one of the classes compared to the others)
 
+## Predictor variable selection
+
+> This section is not required reading for the remainder of the textbook.
+
+Another potentially important part of tuning your classifier is to choose which
+variables from your data will be treated as predictor variables. Based
+on your initial exploratory analysis, you can choose
+anything from using a single predictor variable to using every variable in your
+data; the K-nearest neighbours algorithm accepts any number of
+predictors. However, it is **not** the case that using more predictors always
+yields better predictions! In fact, including irrelevant predictors can
+actually negatively affect classifier performance.
+
+Let's take a look at an example where K-nearest neighbours performs
+worse when given more predictors to work with. Here we have modified
+the breast cancer data to include the variables `Smoothness` and `Concavity`,
+as well as 500 irrelevant variables
+named `Irrelevant1` to `Irrelevant500`, which
+we generated ourselves using a random number generator.
+You will see from the scatter plot below with two example
+variables `Irrelevant1`, `Irrelevant2` that
+these irrelevant variables have no meaningful relationship
+with the `Class` response variable.
+<center>
+```{r 06-irrelevant-loaddata, echo = FALSE, warning = FALSE, fig.cap = "A scatter showing two of the synthetic variables that were added to the data", fig.retina = 2, out.width = "600"}
+cancer_irrelevant <- read_csv("data/wdbc_irrelevant.csv") %>%
+                mutate(Class = as_factor(Class))
+
+plot_irrelev <- cancer_irrelevant %>%
+  ggplot(aes(x = Irrelevant1, y = Irrelevant2, color = Class)) +
+  geom_point(alpha = 0.5) +
+  labs(color = "Diagnosis") +
+  scale_color_manual(labels = c("Malignant", "Benign"), values = cbPalette)
+
+plot_irrelev
+```
+</center>
+
+Next, we build a sequence of K-NN classifiers that include both `Smoothness` and `Concavity` as predictor variables,
+but also increasingly many irrelevant variables. Each model is tuned via 5-fold cross-validation.
+We show the plot of estimated accuracy versus the number of irrelevant predictors below.
```

Comment:
```suggestion
We show the plot of estimated cross-validation accuracy versus the number of irrelevant predictors below.
```

Reply:
Also, we should reference the figures in the text as we refer to them (instead of saying below).

Reply:
This is something we should definitely address throughout the whole book -- I will open an issue to remind us to do it

Reply:
Issue #166